### PR TITLE
Frontend Refactor: Streaming tags

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Moves all tags endpoints over to the new frontend async pipeline. Additionally adds:

- Streaming support to all tags endpoints
- Cache support for tags 
  - See notes. I'm wondering if caching tag values is worth it.
- Integration tests for streaming tags endpoints
- CLI support for tags endpoints

**Which issue(s) this PR fixes**:
Work toward #3400 

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/frontend/cache_keys.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/grafana/tempo/tempodb/backend"
+)
+
+const (
+	cacheKeyPrefixSearchJob       = "sj:"
+	cacheKeyPrefixSearchTag       = "st:"
+	cacheKeyPrefixSearchTagValues = "st:"
```
Comment (mapno) [inline modules/frontend/cache_keys.go:13]: Same prefix, is that correct?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ## main / unreleased
 
+* [FEATURE] Added gRPC streaming endpoints for all tag queries. [#3460](https://github.com/grafana/tempo/pull/3460) (@mapno)
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply CHANGELOG.md:3]: Beets

[the remark this answers is not in the document]
Comment (joe-elliott) [reply CHANGELOG.md:3]: Battlestar Galactica

[the remark this answers is not in the document]
Comment (joe-elliott) [reply CHANGELOG.md:3]: Small Dogs

[the remark this answers is not in the document]
Comment (joe-elliott) [reply CHANGELOG.md:3]: Paella

[the remark this answers is not in the document]
Comment (joe-elliott) [reply CHANGELOG.md:3]: Spain

[the remark this answers is not in the document]
Comment (joe-elliott) [reply CHANGELOG.md:3]: Fixed!

File: cmd/tempo-cli/cmd-query-search-tag-values.go [hunk 3]
```diff
@@ -1,23 +1,115 @@
...
+	"time"
+
+	"github.com/grafana/dskit/user"
 	"github.com/grafana/tempo/pkg/httpclient"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 )
 
 type querySearchTagValuesCmd struct {
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply cmd/tempo-cli/cmd-query-search-tag-values.go:17]: added! (for gRPC)

http requires typing more things

File: integration/e2e/config-all-in-one-local.yaml [hunk 4]
```diff
@@ -1,4 +1,5 @@
 target: all
+stream_over_http_enabled: true
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply integration/e2e/config-all-in-one-local.yaml:2]: streaming in the all in one is done using the shared port:

https://github.com/grafana/tempo/pull/3460/files#diff-0a565c64b055da77723873043e49f4e2b89ed84165f78320ee9e4508ee376857R323

File: modules/frontend/tag_handlers_test.go [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+	httpResp := httptest.NewRecorder()
+
+	ctx, cancel := context.WithCancel(httpReq.Context())
+	ctx = user.InjectOrgID(ctx, "tenant")
+
+	httpReq = httpReq.WithContext(ctx)
+	go func() {
+		time.Sleep(50 * time.Millisecond)
+		cancel()
+	}()
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply modules/frontend/tag_handlers_test.go:115]: these tests are comprehensive, but poorly organized. as i move more endpoints to the new streaming pipeline i want to clean them up but they are becoming quite unruly.

in this case:
1) we put the http request in the http pipeline here: https://github.com/grafana/tempo/pull/3460/files#diff-512f96e1e3fb3f728660709269d6e2089a469dd022967684aaea6d035a2c4203R117
2) the go func you've highlighted cancels the context 50ms later
3) 50ms beats the timeout here: https://github.com/grafana/tempo/blob/main/modules/frontend/search_handlers_test.go#L66
4) so we expect a context canceled error (which we get)

we need the async go func so we cancel the context while the http request is in flight.

File: modules/frontend/cache_keys.go (hunk 1 above)
Comment (joe-elliott) [reply modules/frontend/cache_keys.go:13]: nope!

**Review by mapno — APPROVED** [review]
LGTM

File: docs/sources/tempo/api_docs/_index.md [hunk 6]
```diff
@@ -631,52 +631,14 @@ To enable the streaming service over the HTTP port for use with Grafana, set the
 stream_over_http_enabled: true
 ```
 
-The below `rpc` call returns only traces that are new or have updated each time `SearchResponse` is returned except for the last response.
-The final response sent is guaranteed to have the entire resultset.
+The query frontend supports the following interface. For complete details of all objects [refer here](https://github.com/grafana/tempo/blob/main/pkg/tempopb/tempo.proto)
```
Comment (knylander-grafana) [inline docs/sources/tempo/api_docs/_index.md:634]:
```suggestion
The query frontend supports the following interface. Refer to [`tempo.proto`](https://github.com/grafana/tempo/blob/main/pkg/tempopb/tempo.proto) for complete details of all objects.
```

File: docs/sources/tempo/operations/tempo_cli.md [hunk 7]
```diff
@@ -99,6 +99,46 @@ Options:
...
+Arguments:
+- `host-port` A host/port combination for Tempo. The scheme will be inferred based on the options provided.
+- `tag` The fully qualified traceql tag to search for. e.g. `resource.service.name`
+- `start` Start of the time range to search: (YYYY-MM-DDThh:mm:ss)
+- `end` End of the time range to search: (YYYY-MM-DDThh:mm:ss)
+
+Options:
+- `--org-id <value>`      Organization ID (for use in multi-tenant setup).
+- `--use-grpc`            Use GRPC streaming
+- `--path-prefix <value>` String to prefix search paths with
```
Comment (knylander-grafana) [inline docs/sources/tempo/operations/tempo_cli.md:137]: This is the kind of thing you might consider using a [definition list](https://www.markdownguide.org/extended-syntax/#definition-lists), which is supported by Hugo. That the top item (the term) is bold, and the definition appears underneath. 

```suggestion
Arguments:
`host-port` 
: A host/port combination for Tempo. The scheme will be inferred based on the options provided.

`tag` 
: The fully qualified traceql tag to search for; for example: `resource.service.name`

`start`
: Start of the time range to search: (`YYYY-MM-DDThh:mm:ss`)

`end`
: End of the time range to search: (`YYYY-MM-DDThh:mm:ss`)

Options:
`--org-id <value>`
: Organization ID (for use in multi-tenant setup).

`--use-grpc`
: Use GRPC streaming

`--path-prefix <value>`
: String to prefix search paths with
```

**Review by knylander-grafana — COMMENTED** [review]
Thank you for updating the docs! I had a minor formatting suggestion.

**Pull request merged**